numpyro/util.py

```diff
@@ -309,3 +311,227 @@ def soft_vmap(fn, xs, batch_ndims=1, chunk_size=None):
     map_ndims = int(num_chunks > 1) + int(chunk_size > 1)
     ys = tree_map(lambda y: jnp.reshape(y, (-1,) + jnp.shape(y)[map_ndims:])[:batch_size], ys)
     return tree_map(lambda y: jnp.reshape(y, batch_shape + jnp.shape(y)[1:]), ys)
+
+
+def get_model_relations(model, model_args=None, model_kwargs=None, num_tries=10):
+    """
+    Infer relations of RVs and plates from given model and optionally data.
+    See issue #949 on pyro-ppl/numpyro for more details.
+
+    :param int num_tries: times to trace model to detect discrete -> continuous dependency.
+    """
+    # TODO: put import in more sensible location
+    from numpyro import handlers
+
+    model_args = model_args or ()
+    model_kwargs = model_kwargs or {}
+
+    trace = handlers.trace(handlers.seed(model, 0)).get_trace(*model_args, **model_kwargs)
+    obs_sites = [name for name, site in trace.items()
+                 if site['type'] == 'sample' and site['is_observed']]
+
```

Comment:
@kpj This will give you names of the distributions
```python
import numpyro.distributions as dist

def _get_dist_name(fn):
    if isinstance(fn, (dist.Independent, dist.ExpandedDistribution, dist.MaskedDistribution)):
        return _get_dist_name(fn.base_dist)
    return type(fn).__name__

sample_dist = {name: _get_dist_name(site["fn"]) for name, site in trace.items() if site["type"] == "sample"}
```

Reply:
Awesome! Was planning to do exactly that :-)

And because positioning the text node and aligning the text properly seems difficult, I'll leave it off by default.